File: julia/02-variables/variables.jl
# Variable assignment - no keywords needed
x = 42
pi_approx = 3.14159
name = "Julia"
is_fast = true

# Check types with typeof()
println("x: $x ($(typeof(x)))")
println("pi_approx: $pi_approx ($(typeof(pi_approx)))")
println("name: $name ($(typeof(name)))")
println("is_fast: $is_fast ($(typeof(is_fast)))")

println()

# Integer types - Julia chooses Int64 on 64-bit systems
a = 10              # Int64
b = 0xff            # UInt8 (hexadecimal literal)
c = 0b1010          # UInt8 (binary literal)
big_num = 10_000_000  # Underscores for readability

println("a: $a ($(typeof(a)))")
println("b: $b ($(typeof(b)))")
println("c: $c ($(typeof(c)))")
println("big_num: $big_num ($(typeof(big_num)))")

println()

# Floating-point types
f1 = 1.0            # Float64 (default)
f2 = 1.0f0          # Float32 (note the 'f' suffix)
f3 = 2.5e3          # Scientific notation: 2500.0

println("f1: $f1 ($(typeof(f1)))")
println("f2: $f2 ($(typeof(f2)))")
println("f3: $f3 ($(typeof(f3)))")

println()

# Special numeric values
println("Infinity: $(Inf)")
println("Negative infinity: $(-Inf)")
println("Not a number: $(NaN)")
println("Machine epsilon: $(eps(Float64))")

println()

# Characters vs Strings
ch = 'A'             # Char (single quotes)
str = "A"            # String (double quotes)
println("'A' is $(typeof(ch)), \"A\" is $(typeof(str))")

println()

# Nothing - Julia's null equivalent
result = nothing
println("nothing: $result ($(typeof(result)))")

File: matlab/02-variables/variables.m
% Variable assignment - no keywords needed
x = 42;
pi_approx = 3.14159;
name = 'MATLAB';
is_ready = true;

% Check types with class()
fprintf('x: %d (%s)\n', x, class(x));
fprintf('pi_approx: %.5f (%s)\n', pi_approx, class(pi_approx));
fprintf('name: %s (%s)\n', name, class(name));
fprintf('is_ready: %d (%s)\n', is_ready, class(is_ready));

fprintf('\n');

% Numeric types - doubles are the default
a = 10;                 % double (default for all numeric literals)
b = int32(10);          % explicit 32-bit integer
c = uint8(255);         % unsigned 8-bit integer
d = single(3.14);       % 32-bit float

fprintf('=== Numeric Types ===\n');
fprintf('a = %d (%s) - default is double\n', a, class(a));
fprintf('b = %d (%s)\n', b, class(b));
fprintf('c = %d (%s)\n', c, class(c));
fprintf('d = %.2f (%s)\n', d, class(d));

fprintf('\n');

% Special numeric values
fprintf('=== Special Values ===\n');
fprintf('Infinity: %f\n', Inf);
fprintf('Negative infinity: %f\n', -Inf);
fprintf('Not a number: %f\n', NaN);
fprintf('Machine epsilon: %.15e\n', eps);
fprintf('Largest double: %.6e\n', realmax);
fprintf('Smallest positive double: %.6e\n', realmin);

fprintf('\n');

% Complex numbers - built-in support
z1 = 3 + 4i;
z2 = complex(1, -2);

fprintf('=== Complex Numbers ===\n');
fprintf('z1 = %s\n', num2str(z1));
fprintf('z2 = %s\n', num2str(z2));
fprintf('Real part of z1: %.1f\n', real(z1));
fprintf('Imaginary part of z1: %.1f\n', imag(z1));
fprintf('Magnitude of z1: %.1f\n', abs(z1));

fprintf('\n');

% Logical (boolean) values
fprintf('=== Logical Values ===\n');
flag1 = true;
flag2 = false;
flag3 = logical(1);
fprintf('true: %d (%s)\n', flag1, class(flag1));
fprintf('false: %d (%s)\n', flag2, class(flag2));
fprintf('logical(1): %d (%s)\n', flag3, class(flag3));
fprintf('5 > 3 produces: %d (%s)\n', 5 > 3, class(5 > 3));
